### PR TITLE
Group membership api implementation

### PR DESCRIPTION
**Pull request by thekemkid**
This handles auto assigning partitions, etc, when joining a group

### CODE REVIEW

**Comment by mcdonnelldean** [comment]
Wow @thekemkid This looks much better 💃 

@pelger I need writes across the board to review, turn on travis, etc.

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+sudo: false
+language: node_js
+node_js:
+  - "0.10"
+  - "0.12"
+  - "iojs-v1"
```
Comment (mcdonnelldean) [inline .travis.yml]: We can drop IOJS at this point?

File: LICENSE [hunk 2]
```diff
@@ -1,4 +1,4 @@
-The MIT License
+The MIT License (MIT)
```
Comment (mcdonnelldean) [inline LICENSE]: Good spot. Change the copyright below to 2013 - 2016 Peter Elger and other contributors

File: README.md [hunk 3]
```diff
@@ -1,17 +1,14 @@
-# kafkaesque
+# kafkaesque [![Build Status](https://travis-ci.org/thekemkid/Kafkaesque.svg?branch=group-membership)](https://travis-ci.org/thekemkid/Kafkaesque)
```
Comment (mcdonnelldean) [inline README.md]: Use named links, put link at the bottom

File: README.md [hunk 4]
```diff
@@ -1,17 +1,14 @@
...
 
-## A Node.js Kafka client
-kafkaesque is a node.js client for Apache Kafka supporting upwards of v0.8 of the Kafka protocol only. Kafkaesque does not require any connection to zookeeper, rather it uses the kafka metadata protocol request to determine how it should best connect to the cluster. You need only provide Kafkaesque with the details of a single broker in any Kafka cluster and it will figure out the rest.
-
-The current 0.8 release of Kafka does not appear to support the full protocol set as described here: [https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol](https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol). Specifically the offset/commit/fetch API.
+[![NPM](https://nodei.co/npm/kafkaesque.png?downloads=true&downloadRank=true&stars=true)](https://nodei.co/npm/kafkaesque/)
```
Comment (mcdonnelldean) [inline README.md]: Named links

File: README.md [hunk 5]
```diff
@@ -1,17 +1,14 @@
...
 
-Kafkaesque will uses API as opposed to reading meta commit information from zookeeper when it is full supported in Kafka.
+## A Node.js Kafka client
+kafkaesque is a node.js client for Apache Kafka. This client supports v0.8.1 and upwards of the Kafka protocol. Kafkaesque does not require any connection to zookeeper as it uses the kafka [wire protocol](https://cwiki.apache.org/confluence/display/KAFKA/A+Guide+To+The+Kafka+Protocol) to determine how it should best connect and manage connections to the cluster. You need only provide Kafkaesque with the details of a single broker in any Kafka cluster and it will figure out the rest.
```
Comment (mcdonnelldean) [inline README.md]: Named links

File: README.md [hunk 6]
```diff
@@ -21,111 +18,212 @@ npm install kafkaesque
...
+    console.log(error);
   });
 });
 ```
 
+## API
+
+### Configuration:
+```javascript
+var kafkaesque = require('kafkaesque')(configObj);
```
Comment (mcdonnelldean) [inline README.md]: Nesting spacing seems very deep here? 2 vs 4?

File: README.md [hunk 7]
```diff
@@ -21,111 +18,212 @@ npm install kafkaesque
...
-	* params.topic - the topic name, required
-    * params.partition - the partition id, required
-    * offset - the starting offset, if unspecified kafkaesque uses the latest commmited position against this topic / partition pair
-    *  maxWait - the maximum poll wait time, if unspecified defaults to 5 seconds
-    *   minBytes - the minimum bytes that should be available before returning, if unspecified defaults to 50 bytes
+We hope that this code is useful, please feel free to get in touch if you need help or support: @pelger or @thekemkid.
 
-## Support
-Hope that this code is useful, please feel free to get in touch if you need help or support: @pelger
+## License
```
Comment (mcdonnelldean) [inline README.md]: 2013 - 2016

File: lib/api.js [hunk 8]
```diff
@@ -16,23 +16,36 @@
 
 var assert = require('assert');
 var net = require('net');
+var events = require('events');
 // var hexy = require('hexy');
```
Comment (mcdonnelldean) [inline lib/api.js]: No commented code, remove this line.

File: lib/api.js [hunk 9]
```diff
@@ -46,6 +59,8 @@ module.exports = function(options) {
   var _cbt;
   var _rcvBuf;
 
+  var _groupEvents;
+
 
   var sendMsg = function(msg, correlationId, cb) {
     /*console.log('snd -----------------------------------------------');
```
Comment (mcdonnelldean) [inline lib/api.js]: Remove commented out code.

File: lib/api.js [hunk 10]
```diff
@@ -121,6 +136,7 @@ module.exports = function(options) {
    *   minBytes: the minimum number of bytes that should be available before a response is sent
    */
   var fetch = function(params, cb) {
+    // console.log(params)
```
Comment (mcdonnelldean) [inline lib/api.js]: remove commented out code

File: lib/kafkaesque.js [hunk 11]
```diff
@@ -18,10 +18,15 @@ var assert = require('assert');
 var api = require('./api');
 var events = require('events');
 var _ = require('underscore');
+var reInterval = require('reInterval');
 
-var MAX_WAIT_DEFAULT = 5000;
-var MIN_BYTES_DEFAULT = 2;
 
+// _options = {
```
Comment (mcdonnelldean) [inline lib/kafkaesque.js]: remove commented code

File: lib/kafkaesque.js [hunk 12]
```diff
@@ -114,74 +202,105 @@ module.exports = function(options) {
...
+        });
+      } else {
+        _getTopicMetadata(params, function(err, cluster) {
+          if (err) {
+            return cb(err, null);
+          }
 
-        var leaderKey = _makeBrokerKey(leader);
-        var topicKey = _makeTopicKey(params);
+          // set _topics[_makeTopicKey(params)] = true to cause enclosing
```
Comment (mcdonnelldean) [inline lib/kafkaesque.js]: remove commented code

File: lib/kafkaesque.js (hunk 11 above)
Comment (mcdonnelldean) [inline lib/kafkaesque.js]: Can this whole file be split up better? It's nearly 1000 lines long, very hard to review code like this which means things will slip past. Are there any easy places to break this out?

File: lib/message/request/joinGroup.js [hunk 13]
```diff
@@ -0,0 +1,77 @@
...
+
+  bytes.appendUInt32BE(topics.length);
+  topics.forEach(function (topic) {
+    common.appendString(bytes, topic);
+  });
+
+  // userdata bytes
+  bytes.appendUInt32BE(0);
+  // we could actually put some json in the userdata bytes section using this...
+  // common.appendStringAsBytes(bytes, 'wassssup')
```
Comment (mcdonnelldean) [inline lib/message/request/joinGroup.js]: remove all dead code from this file too

File: lib/message/request/syncGroup.js [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+}
+
+
+var member = function (memberId) {
+  common.appendString(this.buf, memberId);
+
+  return this;
+}
+
+// Example assignments structure...
```
Comment (mcdonnelldean) [inline lib/message/request/syncGroup.js]: This should be in the readme or /docs. Not here

File: lib/message/request/syncGroup.js [hunk 15]
```diff
@@ -0,0 +1,112 @@
...
+//     }
+//   }
+// ]
+var groupAssignment = function (memberAssignments) {
+  var buf = this.buf;
+  buf.appendUInt32BE(memberAssignments.length);
+
+  memberAssignments.forEach(function (member) {
+    common.appendString(buf, member.id);
+    // this.buf.appendUInt16BE(0x0001);
```
Comment (mcdonnelldean) [inline lib/message/request/syncGroup.js]: remove all dead code

File: lib/message/response/joinGroup.js [hunk 16]
```diff
@@ -0,0 +1,93 @@
...
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+ * STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+ * IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+
+'use strict';
+
+var common = require('./common');
+// var hexy = require('hexy');
```
Comment (mcdonnelldean) [inline lib/message/response/joinGroup.js]: Remove dead code

**Comment by mcdonnelldean** [comment]
@thekemkid Ok I've reviewed.

Before I can ok I need to get access to run through travis but it looks very good. Great stuff.

Most of my comments are around removing dead code and splitting up that monster main file. Lets schedule some time for this, this week.

**Comment by thekemkid** [comment]
@mcdonnelldean Thanks for the review! Yeah, it needs to be split better, I'll look into it 👍

File: LICENSE (hunk 2 above)
Comment (thekemkid) [inline LICENSE]: resolved

File: README.md (hunk 6 above)
Comment (thekemkid) [inline README.md]: not sure what you mean here?

File: README.md (hunk 7 above)
Comment (thekemkid) [inline README.md]: done

File: lib/kafkaesque.js (hunk 12 above)
Comment (thekemkid) [inline lib/kafkaesque.js]: This was actually explaining what was going on with the code, so I'm inclined to leave that as is. is that okay?

**Comment by mihaidma** [comment]
great work @thekemkid

**Pull request merged**